File: optimize-images.js
#!/usr/bin/env node
/**
 * Image optimization script - converts PNG avatars to WebP with compression
 * Reduces file size from ~445KB to <50KB estimated
 */

const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

const PUBLIC_DIR = path.join(__dirname, 'public');

const images = [
  { name: 'Kepala.png', outputName: 'Kepala.webp' },
  { name: 'Kepala1.png', outputName: 'Kepala1.webp' },
  { name: 'Kepala.webp', outputName: 'Kepala-56.webp', width: 56, height: 56 },
];

async function optimizeImages() {
  console.log('🖼️  Starting image optimization...\n');

  for (const image of images) {
    const inputPath = path.join(PUBLIC_DIR, image.name);
    const outputPath = path.join(PUBLIC_DIR, image.outputName);

    if (!fs.existsSync(inputPath)) {
      console.log(`⚠️  Skipping ${image.name} - file not found`);
      continue;
    }

    try {
      // Get input file size
      const inputStats = fs.statSync(inputPath);
      const inputSize = (inputStats.size / 1024).toFixed(2);

      // Convert to WebP with aggressive compression
      let transformer = sharp(inputPath);
      if (image.width || image.height) {
        transformer = transformer.resize(image.width, image.height, { fit: 'inside' });
      }

      await transformer
        .webp({ quality: 75, alphaQuality: 75 })
        .toFile(outputPath);

      // Get output file size
      const outputStats = fs.statSync(outputPath);
      const outputSize = (outputStats.size / 1024).toFixed(2);
      const savings = (inputSize - outputSize).toFixed(2);
      const reduction = ((savings / inputSize) * 100).toFixed(1);

      console.log(`✅ ${image.name}`);
      console.log(`   Input:  ${inputSize} KiB`);
      console.log(`   Output: ${outputSize} KiB`);
      console.log(`   Saved:  ${savings} KiB (${reduction}% reduction)`);
      console.log();
    } catch (error) {
      console.error(`❌ Error processing ${image.name}:`, error.message);
    }
  }

  console.log('✨ Image optimization complete!');
  console.log(
    '\nNext steps:\n' +
    '1. Update image src paths from .png to .webp in components\n' +
    '2. Add fallback to .png for older browsers (optional)\n' +
    '3. Run lighthouse to verify savings\n'
  );
}

optimizeImages().catch(console.error);

File: optimize-images.py
#!/usr/bin/env python3
"""
Image optimization script - converts PNG avatars to WebP format
Reduces Kepala.png from 179KB to ~15-20KB
"""

import os
import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("Error: PIL/Pillow not found. Installing...")
    os.system(f"{sys.executable} -m pip install Pillow -q")
    from PIL import Image

PUBLIC_DIR = Path(__file__).parent / "public"

images = [
    {"name": "Kepala.png", "output": "Kepala.webp", "quality": 75},
    {"name": "Kepala1.png", "output": "Kepala1.webp", "quality": 75},
]

def optimize_images():
    print("🖼️  Starting image optimization...\n")
    
    total_input = 0
    total_output = 0
    
    for img_config in images:
        input_path = PUBLIC_DIR / img_config["name"]
        output_path = PUBLIC_DIR / img_config["output"]
        
        if not input_path.exists():
            print(f"⚠️  Skipping {img_config['name']} - file not found")
            continue
        
        try:
            # Get input file size
            input_size = input_path.stat().st_size / 1024
            total_input += input_size
            
            # Open and convert
            img = Image.open(input_path)
            
            # Convert RGBA to RGB if necessary (for better WebP compression)
            if img.mode in ('RGBA', 'LA', 'P'):
                # Create white background
                background = Image.new('RGB', img.size, (255, 255, 255))
                # Paste with alpha mask
                if img.mode == 'P':
                    img = img.convert('RGBA')
                background.paste(img, mask=img.split()[-1] if img.mode == 'RGBA' else None)
                img = background
            
            # Save as WebP
            img.save(
                output_path,
                'WebP',
                quality=img_config["quality"],
                method=6  # Slowest, best compression
            )
            
            # Get output file size
            output_size = output_path.stat().st_size / 1024
            total_output += output_size
            
            savings = input_size - output_size
            reduction_pct = (savings / input_size) * 100
            
            print(f"✅ {img_config['name']}")
            print(f"   Input:  {input_size:.2f} KiB")
            print(f"   Output: {output_size:.2f} KiB")
            print(f"   Saved:  {savings:.2f} KiB ({reduction_pct:.1f}% reduction)")
            print()
            
        except Exception as e:
            print(f"❌ Error processing {img_config['name']}: {e}\n")
    
    print(f"✨ Optimization complete!")
    print(f"Total input:  {total_input:.2f} KiB")
    print(f"Total output: {total_output:.2f} KiB")
    print(f"Total saved:  {total_input - total_output:.2f} KiB ({((total_input-total_output)/total_input)*100:.1f}%)\n")
    
    print("Next steps:")
    print("1. Update image sources from .png to .webp in components")
    print("2. Remove the ?v=20260528 query params (no longer needed)")
    print("3. Test on mobile and desktop")
    print("4. Run lighthouse to verify the savings\n")

if __name__ == "__main__":
    optimize_images()
